nomos-mix/message/src/layered_cipher.rs

```diff
@@ -0,0 +1,331 @@
+///
+/// The max number of layers is limited to the `max_layers` parameter.
+/// Even if the number of data and keys provided for encryption is smaller than `max_layers`,
+/// The cipher always produces the max-size output regardless of the number of data and keys provided,
+/// in order to ensure that all outputs generated by the cipher are the same size.
+///
+/// The cipher also provides the length-preserved decryption.
+/// Even if one layer of encryptions is decrypted, the length of decrypted data is
+/// the same as the length of the original data.
+/// For example:
+///   len(encrypt(k0, [data0, encrypt(k1, [data1])])) == len(encrypt(k1, [data1]))
+pub struct ConsistentLengthLayeredCipher<D: ConsistentLengthLayeredCipherData> {
+    /// All encrypted data produced by the cipher has the same size according to the `max_layers`.
+    pub max_layers: usize,
+    _phantom_data: PhantomData<D>,
+}
+
+pub trait ConsistentLengthLayeredCipherData {
+    // Returns the serialized bytes for an instance of the implementing type
+    fn to_bytes(&self) -> Vec<u8>;
+    // Returns the size of the serialized data. This is a static method.
+    fn size() -> usize;
+}
+
+/// A parameter for one layer of encryption
+pub struct EncryptionParam<D: ConsistentLengthLayeredCipherData> {
+    /// A data to be included in the layer.
+    pub data: D,
+    /// A [`Key`] to encrypt the layer that will include the [`Self::data`].
+    pub key: Key,
+}
+
+/// A set of keys to encrypt/decrypt a single layer.
+pub struct Key {
+    /// A 128-bit key for encryption/decryption
+    pub stream_cipher_key: StreamCipherKey,
+    /// A 128-bit key for computing/verifying integrity MAC
+    pub integrity_mac_key: HeaderIntegrityMacKey,
+}
+
+impl<D: ConsistentLengthLayeredCipherData> ConsistentLengthLayeredCipher<D> {
+    pub fn new(max_layers: usize) -> Self {
+        Self {
+            max_layers,
+            _phantom_data: Default::default(),
+        }
+    }
+
+    /// The total size of fully encrypted output that includes all layers.
+    /// This size is determined by [`D::size`] and [`Self::max_layers`].
+    pub fn total_size(&self) -> usize {
+        Self::single_layer_size() * self.max_layers
+    }
+
+    /// The size of a single layer that contains a data and a MAC.
+    /// The MAC is used to verify integrity of the encrypted next layer.
+    fn single_layer_size() -> usize {
+        D::size() + HEADER_INTEGRITY_MAC_SIZE
+    }
+
+    /// Perform the layered encryption.
+    pub fn encrypt(&self, params: &[EncryptionParam<D>]) -> Result<(Vec<u8>, HeaderIntegrityMac)> {
+        if params.is_empty() || params.len() > self.max_layers {
+            return Err(Error::InvalidParams);
+        }
+
+        params
+            .iter()
+            .take(params.len() - 1) // Exclude the last param that will be treated separately below.
+            .rev() // Data and keys must be used in reverse order to encrypt the inner-most layer first
```

Comment:
Is there a way to make sure this actually happens like that? as caller may have already reverse the order of Params.
Not to do now, but in this case I think we may construct some structure for it. wdyt?